### ISSUE TITLE
Bug: multiple replaceTextRange calls returned in an array do not work unless the order of ranges is totally ordered

### ISSUE THREAD

**Issue by sstchur** [eslint/eslint#16363]
### Environment

Node version: 16.14.2
npm version: 8.5.0
Local ESLint version: 8.13.0
Global ESLint version: N/A
Operating System: Ubuntu

### What parser are you using?

@typescript-eslint/parser

### What did you do?

Full repro available at: https://github.com/sstchur/eslint-repro

Given the following code snippet in a rule test:
```typescript
// A
// B
class Person {
  constructor(A: string, B: string) {}
}
```

And the following (contrived) rule code:
```typescript
return {
  Identifier(node) {
    if (node.name === 'A') {
      context.report({
        node,
        messageId: 'someError',
        fix(fixer) {
          return [
            fixer.replaceTextRange([9,13], '// C'),
            fixer.replaceTextRange([72,73], 'C')
          ]
        }
      });
    }
    else if (node.name === 'B') {
      context.report({
        node,
        messageId: 'someError',
        fix(fixer) {
          return [
            fixer.replaceTextRange([22,26], '// D'),
            fixer.replaceTextRange([82,83], 'D')
          ];
        }
      });
    }
  }
};
```

The fixers for both A and B will run, but only the fixer for A will "stick" (for lack of a better word), despite these ranges all being ranges that do not overlap.

### What did you expect to happen?

I expected the fixers for both A and B to run and the resulting output code to be:
```typescript
// C
// D
class Person {
  constructor(C: string, D: string) {}
}
```

### What actually happened?

Instead, the resulting output is:

```typescript
// C
// B
class Person {
  constructor(C: string, B: string) {}
}
```

Actual ESLint output from running the rule test:
```Difference:

   Expected value to strictly be equal to:
  "
        // C
        // D
        class Person {
          constructor(C: string, D: string) {}
        }"
Received:
  "
        // C
        // B
        class Person {
          constructor(C: string, B: string) {}
        }"

Message:
  Output is incorrect.

Difference:

- Expected
+ Received


          // C
-         // D
+         // B
          class Person {
-           constructor(C: string, D: string) {}
+           constructor(C: string, B: string) {}
          }
```

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

I think it's actually really useful to note that if I change the order of the fixing objects so that they are effectively called in an order in which the ranges are sorted, then everything works, like so:

```typescript
fix(fixer) {
  return [
    fixer.replaceTextRange([9,13], '// C'),
    fixer.replaceTextRange([22,26], '// D'),
  ];
}
...
fix(fixer) {
  return [
    fixer.replaceTextRange([72,73], 'C')
    fixer.replaceTextRange([83,84], 'D')
  ];
}
```

Note that above, the ranges are totally ordered: [9,13], [22,26], [72,73], [83,84].

The docs clearly call out the fixing objects cannot overlap.  However, it says (and I quote):
"If you make a fix() function which returns multiple fixing objects, those fixing objects must not be overlapped."

My interpretation of that is that the fixing objects that are part of an array from within a `fix` function must not overlap (which the ones in my example do not).

However, it seems like subsequent calls to a fixer cannot "go backwards" (so to speak).  I'm not sure if the comment in the docs about "fixing object must not be overlapped" was meant to cover this case?  I didn't interpret it as such, but maybe I misunderstood.

In any case, a more fundamental question: should it be possible to calling `replaceTextRange([x,y], 'whatever')` and then later call `replaceTextRange([x-something, y-something], 'blah')` (so long as the ranges `[x,y]` and `[x-something,y-something]` are NON overlapping?  Or **must** `replaceTextRange([x-something, y-something], 'blah')` be called **before** `replaceTextRange([x,y], 'whatever')` (as appears to be the case today)?

**Comment by sstchur** [comment]
Actually, a clarification:  It seems that the ranges don't have to be totally ordered within a single `context.report` call; it's just that all the ranges in one `context.report`, have to come before all the ranges in subsequent `context.report` calls.

In addition to the order I mentioned in the "Additional comments" section above, the following order would also work:

```typescript
fix(fixer) {
      return [
        fixer.replaceTextRange([22,26], '// D'),
        fixer.replaceTextRange([9,13], '// C')
        
      ]
    }
  });
}
...
fix(fixer) {
  return [
    
    fixer.replaceTextRange([83,84], 'D'),
    fixer.replaceTextRange([72,73], 'C'),
  ];
} 
```

Note that in this case, the ranges are not totally ordered, but since A's fixer runs first, [22,26] and [9,13] both run before [83,84] and [72,73], and that appears to work as well (even though, within each fixer, the ranges are not ordered).

**Comment by sstchur** [comment]
UPDATE: I wonder if this is actually a RuleTester issue?  I started debugging through `report-translator.js`, specifically the function `mergeFixes`, and from what I can tell, both fixes get applied properly.  That had me wondering, so I created a fake project and enabled my fake rule in that project.  I allowed VSCode to squiggly underline the offending code, and I chose "Quick fix" -> "Fix this problem".  It actually worked.  It updated both places in the code (code and comment), and when I chose "Quick fix" a second time for the second line of offending code, it again updated both places in the code (code and comment).

As a final test, I reverted the changes, and used the eslint cli to run `yarn lint --fix`, and again, it seemed to work correctly!

So now I'm wondering if there might be a bug/issue w/ the RuleTester, but ESLint itself seems like it might be working as intended.

**Comment by mdjermanovic** [comment]
`--fix` runs up to 10 passes of autofixing, while RuleTester by design runs only one.

This works as intended because fixes with overlapping ranges cannot be applied in the same pass, and these two fixes do overlap.

When you call `context.report()`, ESLint translates provided object to a [LintMessage](https://eslint.org/docs/latest/developer-guide/nodejs-api#-lintmessage-type) object. `fix()` is run to calculate the value of the `fix` property, which should be an [EditInfo](https://eslint.org/docs/latest/developer-guide/nodejs-api#-editinfo-type) object containing a single range and replacement text. When `fix()` returns an array (with ranges that don't need to be ordered, but must not overlap), ESLint translates it into one continuous range that starts with the leftmost and ends with the rightmost index of the provided ranges, and any missing segments are filled in with the source code text.

In your original example, the rule produces two lint messages. The first message has fix range `[9, 73]`, and the second has fix range `[22, 83]`, so they are overlapping and RuleTester will apply only the first one. You can run ESLint with `-f json` (and without `--fix`) to see the replacement text.

In the example from https://github.com/eslint/eslint/issues/16363#issuecomment-1261476493, the ranges are `[9, 22]` and `[72, 84`], so the two fixes can be applied in the same pass.

**Comment by sstchur** [comment]
Thanks for the explanation!
I totally understand what you're saying about the RuleTester and ranges `[9,73]` and `[22,83]`, which do indeed overlap.  When I debugged the ESLint code, I saw this in action, so it makes perfect sense to me.  But my point is that the docs aren't clear about this.  They seem to imply that as long as a single call to `context.report` specifies non-overlapping fixes in the returned array, you're good to go.  And in a sense, this seems to be true, since it does work in "the real world" even though it fails in the test.

Is it not worth clarifying somewhere in the docs that the RuleTester only does a single pass by design, and as such, scenarios like the once I described will not be testable?  Or perhaps that's buried in the docs somewhere, and I just missed it?

**Comment by mdjermanovic** [comment]
> They seem to imply that as long as a single call to `context.report` specifies non-overlapping fixes in the returned array, you're good to go. And in a sense, this seems to be true, since it does work in "the real world" even though it fails in the test.

Yes, rules generally shouldn't be concerned about overlapping fixes. as `--fix` runs linting and fixing again. 

> Is it not worth clarifying somewhere in the docs that the RuleTester only does a single pass by design, and as such, scenarios like the once I described will not be testable?

[RulTester docs](https://eslint.org/docs/latest/developer-guide/nodejs-api#ruletester) do say that there is only a single pass:

> * `output` (string, required if the rule fixes code): Asserts the output that will be produced when using this rule for a single pass of autofixing (e.g. with the `--fix` command line flag). If this is `null`, asserts that none of the reported problems suggest autofixes.

but that `` e.g. with the `--fix` command line flag `` part is confusing because `--fix` is multi-pass. PR to clarify this is welcome!

**Comment by sstchur** [comment]
Sounds good. Thanks for your help and explanation.  I'll see if I can find some time to put together a simple PR to clarify this a little.
